README.md
# Metricky
Display metrics about your database in your application. Depends on the awesome [Chartkick](https://github.com/ankane/chartkick) and [Groupdate](https://github.com/ankane/groupdate).

## Demo (with code!)

[metricky-demo.herokuapp.com](https://metricky-demo.herokuapp.com)

## Usage

Make this in Ruby:

<img src="https://i.imgur.com/PQhFyAE.png" alt="Metricky example">

## Generate it 

`rails g metricky:metric TotalUsersMetric --scope User --type :count --period :day`

## Display it 

In your view where you want to display the metric: 

```erbruby
render_metric :total_users
```

## Customize it 

In `app/metrics/total_users_metric.rb`

```ruby 
class TotalUsersMetric < ApplicationMetric 
  def scope
    User
  end

  def type
    :count
  end

  def period
    :day 
  end
end
```

## Installation
Add this line to your application's Gemfile:

```ruby
gem 'metricky'
```

And then execute:
```bash
$ bundle
```

Or install it yourself as:
```bash
$ gem install metricky
```

Then drop in Chartkick into your `application.js` (or similar):

```javascript
//= require chartkick
```

## Customizing

~~Blatantly ripped from~~ Super-inspired by Laravel Nova's metric system.

### Sending more data into the metric

Need to pass `current_user`? 

```erbruby
render_metric :users, user: current_user, account: current_account
```

`user` and `account` will be `attr_reader`-ized.

### Value to calculate

In your metric, define columns:

```ruby 
def columns
  :id 
end
```

### Grouping the value

To `User.all.group(:color).count`

```ruby
def group
  :color 
end
```

### Grouping by period (day, month, year, quarter, etc.)

In your metric, define what period:

```ruby
def period 
  :day 
end
```

This can be any one of `Groupdate::PERIODS`

Define what column should be used:

```ruby
def period_column
  :created_at 
end
```

### Value type 

In your metric, define what type:

```ruby
def type 
  :count 
end
```

This can be any one of `:min, :max, :sum, :count, :average`

### Ranges

Ranges are what values are available on the form (used to query the metric, if applicable) via the `range_column`

```ruby 
def range_column
  :created_at
end
```

Defaults are `all`, `today`, `24 hours`, `3 days`, `7 days`, `10 days`, `14 days`, `30 days`, `3 months`, `6 months`, `12 months`

#### Creating a new range
In your metric, define what ranges as a class method:

```ruby
class TotalUsersMetric < ApplicationMetric
  register_range '15h', label: "15 hours" do 
    15.hours.ago   
  end
end
```

#### Removing a range

```ruby
class TotalUsersMetric < ApplicationMetric
  remove_ranges '24h', '7d' # an array 
  remove_range '3d' # individual 
end
```

#### Removing all ranges

```ruby
class TotalUsersMetric < ApplicationMetric
  reset_ranges!
end
```

#### Setting the default range 

```ruby
class TotalUsersMetric < ApplicationMetric
  default_range '24h'
end
```

#### Customizing the label

Use `collection_label`

```ruby 
class TotalUsersMetric < ApplicationMetric
  def collection_label(range_thing)
    "Born #{range_thing.value.call}"
  end
end 
```

Need helpers? `h`

```
class TotalUsersMetric < ApplicationMetric
  def collection_label(range_thing)
    "Born #{h.time_ago_in_words(range_thing.value.call)}"
  end
end 
```

### Partial

In your metric, define the partial path:

```ruby 
def to_partial_path
  "shared/metric"
end
```

Take a look at `app/views/metricky/_metric.html.erb` for implementation.

## Contributing
Add stuff.

## License
The gem is available as open source under the terms of the [MIT License](https://opensource.org/licenses/MIT).


CHANGELOG.md
# 0.9.0
* Performance improvements.

## 0.8.0
* Mountable controller for rendering charts asynchronously
* That's a hard word
* Add `json?` to return json and subsequently render the chart via an ajax request
* Add `chart?` 

## 0.7.6
* Add `group`

## 0.7.5
* Add `format_value(value)` to format the resulting value

## 0.7.4
* Change the default `form?` to reflect if there are any ranges
* `remove_range` now deletes the range from the `ranges` hash instead of adding the value to an array of keys which were excluded from ranges
* `reset_ranges!` to remove all ranges

## 0.7.3
* Made `form?` a public method

## 0.7.2
* Added `form?` to enable/disable the form

## 0.1.0 ... 0.7.1 ..
* Nobody used this

lib/metricky/ranges.rb
module Metricky
  class Base
    class_attribute :ranges, default: {}
    class_attribute :default_range_key, default: 'all'

    def self.default_range(val = nil)
      self.default_range_key = val
    end

    # Rewrite the priority of a range thing.
    #
    # class UserMetric
    #   range_priority '24h', 99
    # end
    def self.range_priority(key, priority)
      self.ranges[key.to_sym].priority = priority
    end

    # Register a range. Priority is used for the select order.
    #
    # @param [String] key The value for the option in the select for the form
    # @param [String] label The text shown on the option in the select for the form
    # @param [Integer] priority What order the resulting collection
    # @param [Block/Proc] block The Ruby-converted value. Usually a DateTime/Date/Time object.
    #
    # class UserMetric
    #   register_range '15w', label: '15 weeks', priority: 99 do
    #     15.weeks.ago
    #   end
    # end
    #
    # @return RangeThing
    def self.register_range(key, label: nil, priority: nil, &block)
      label ||= key
      priority ||= self.ranges.size
      self.ranges[key.to_sym] = RangeThing.new(label, priority, block)
    end

    # Removes the listed ranges from the select
    def self.remove_ranges(*keys)
      keys.each { self.remove_range(key) }
    end

    def self.remove_range(key)
      self.ranges.delete(key.to_sym)
    end

    # Remove all ranges
    def self.reset_ranges!
      self.ranges = {}
      self.default_range_key = nil
    end

    register_range 'all', label: 'All' do
      nil
    end

    register_range 'today', label: 'Today' do
      DateTime.now.beginning_of_day
    end

    register_range '24h', label: '24 hours' do
      24.hours.ago
    end

    [3, 7, 10, 14, 21, 30].each do |day|
      register_range "#{day}d", label: "#{day} Days" do
        day.days.ago
      end
    end

    register_range "MTD", label: "MTD" do
      DateTime.now.beginning_of_month
    end

    register_range "QTD", label: "QTD" do
      DateTime.now.beginning_of_quarter
    end

    register_range "YTD", label: "YTD" do
      DateTime.now.beginning_of_year
    end

    # Lookup the passed range and convert it to a value for our ActiveRecord query
    def range_to_value
      return nil if range.nil?
      if val = self.ranges[range.to_sym]
        val.value.call
      else
        raise TypeError, "invalid range #{range}. Please define it."
      end
    end
    deprecate_and_alias_method :range_to_value, :range_value

    # What column to specify for the range calculation. Normally `created_at`
    def range_column
      'created_at'
    end

    def collection_label(range_thing)
      "#{range_thing.label}"
    end

    def range
      params.dig(form_name, :range) || default_range_key
    end

    # Used in the HTML form for the metric as the select options
    def range_collection
      ranges.sort_by { |_, range| range.priority }.collect { |key, range_thing| [collection_label(range_thing), key] }
    end
  end
end